File: placement-flutter/storyly_placement_flutter/ios/Classes/Common/Data/Widgets/StoryBarData.swift
import Foundation
import StorylyCore
import StorylyPlacement
import StorylyStoryBar

// MARK: - StoryBar Data Payload Encoding

func encodeStoryBarDataPayload(_ data: StoryBarDataPayload) -> [String: Any] {
    return [
        "type": STRWidgetType.storyBar.description,
        "items": data.items.compactMap { encodeStoryGroupItem($0) }
    ]
}

func encodeStoryBarPayload(_ payload: StoryBarPayload?) -> [String: Any]? {
    guard let payload = payload else { return nil }
    
    return [
        "component": encodeStoryBarComponent(payload.component),
        "group": encodeStoryGroupItem(payload.group),
        "story": encodeStory(payload.story)
    ].compactMapValues { $0 }
}

// MARK: - Story Group Encoding

func encodeStoryGroupItem(_ group: StoryGroup?) -> [String: Any]? {
    guard let group = group else { return nil }
    
    return ([
        "id": group.uniqueId,
        "title": group.title,
        "name": group.name,
        "index": group.index,
        "seen": group.seen,
        "iconUrl": group.iconUrl?.absoluteString,
        "iconVideoUrl": group.iconVideoUrl?.absoluteString,
        "iconVideoThumbnailUrl": group.iconVideoThumbnailUrl?.absoluteString,
        "pinned": group.pinned,
        "type": group.type.description,
        "style": encodeStoryGroupStyle(group.style),
        "nudge": group.nudge,
        "stories": group.stories.compactMap { encodeStory($0) }
    ] as [String: Any?]).compactMapValues { $0 }
}

func encodeStoryGroupStyle(_ style: StoryGroupStyle?) -> [String: Any]? {
    guard let style = style else { return nil }
    
    return ([
        "badge": encodeStoryGroupBadgeStyle(style.badge),
        "borderUnseenColors": style.borderUnseenColors?.map { $0.toHexString() },
        "textUnseenColor": style.textUnseenColor?.toHexString()
    ] as [String: Any?]).compactMapValues { $0 }
}

func encodeStoryGroupBadgeStyle(_ badgeStyle: StoryGroupBadgeStyle?) -> [String: Any]? {
    guard let badgeStyle = badgeStyle else { return nil }
    
    return ([
        "text": badgeStyle.text,
        "textColor": badgeStyle.textColor?.toHexString(),
        "backgroundColor": badgeStyle.backgroundColor?.toHexString(),
        "endTime": badgeStyle.endTime,
        "template": badgeStyle.template
    ] as [String: Any?]).compactMapValues { $0 }
}

// MARK: - Story Encoding

func encodeStory(_ story: Story?) -> [String: Any]? {
    guard let story = story else { return nil }
    
    return ([
        "id": story.uniqueId,
        "title": story.title,
        "name": story.name,
        "index": story.index,
        "seen": story.seen,
        "previewUrl": story.previewUrl?.absoluteString,
        "actionUrl": story.actionUrl,
        "actionProducts": story.actionProducts?.map { encodeSTRProductItem($0) },
        "currentTime": story.currentTime,
        "storyComponentList": story.storyComponentList?.compactMap { encodeStoryBarComponent($0) }
    ] as [String: Any?]).compactMapValues { $0 }
}

// MARK: - Story Bar Component Encoding

func encodeStoryBarComponent(_ component: StoryBarComponent?) -> [String: Any]? {
    guard let component = component else { return nil }
    
    var result: [String: Any] = [
        "id": component.id,
        "type": component.type.get(),
        "customPayload": component.customPayload as Any
    ]
    
    // Add component-specific fields
    switch component {
    case let quiz as StoryQuizComponent:
        result["title"] = quiz.title
        result["options"] = quiz.options
        result["rightAnswerIndex"] = quiz.rightAnswerIndex
        result["selectedOptionIndex"] = quiz.selectedOptionIndex
        
    case let imageQuiz as StoryImageQuizComponent:
        result["title"] = imageQuiz.title
        result["options"] = imageQuiz.options
        result["rightAnswerIndex"] = imageQuiz.rightAnswerIndex
        result["selectedOptionIndex"] = imageQuiz.selectedOptionIndex
        
    case let poll as StoryPollComponent:
        result["title"] = poll.title
        result["options"] = poll.options
        result["selectedOptionIndex"] = poll.selectedOptionIndex
        
    case let emoji as StoryEmojiComponent:
        result["emojiCodes"] = emoji.emojiCodes
        result["selectedEmojiIndex"] = emoji.selectedEmojiIndex
        
    case let rating as StoryRatingComponent:
        result["emojiCode"] = rating.emojiCode
        result["rating"] = rating.rating
        
    case let promoCode as StoryPromoCodeComponent:
        result["text"] = promoCode.text
        
    case let comment as StoryCommentComponent:
        result["text"] = comment.text
        
    case let swipe as StorySwipeComponent:
        result["text"] = swipe.text
        result["actionUrl"] = swipe.actionUrl
        result["products"] = swipe.products?.map { encodeSTRProductItem($0) }
        
    case let button as StoryButtonComponent:
        result["text"] = button.text
        result["actionUrl"] = button.actionUrl
        result["products"] = button.products?.map { encodeSTRProductItem($0) }
        
    case let productTag as StoryProductTagComponent:
        result["actionUrl"] = productTag.actionUrl
        result["products"] = productTag.products?.map { encodeSTRProductItem($0) }
        
    case let productCard as StoryProductCardComponent:
        result["text"] = productCard.text
        result["actionUrl"] = productCard.actionUrl
        result["products"] = productCard.products?.map { encodeSTRProductItem($0) }
        
    case let productCatalog as StoryProductCatalogComponent:
        result["actionUrlList"] = productCatalog.actionUrlList
        result["products"] = productCatalog.products?.map { encodeSTRProductItem($0) }
        
    default:
        break
    }
    
    return result.compactMapValues { $0 }
}

File: flutter/storyly_placement_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_placement_flutter/common/data/widgets/StoryBarData.kt
package com.appsamurai.storyly.storyly_placement_flutter.common.data.widgets

import com.appsamurai.storyly.core.STRWidgetType
import com.appsamurai.storyly.storybar.data.StoryBarDataPayload
import com.appsamurai.storyly.storybar.data.model.StoryBarPayload
import com.appsamurai.storyly.storybar.ui.model.Story
import com.appsamurai.storyly.storybar.ui.model.StoryBarComponent
import com.appsamurai.storyly.storybar.ui.model.StoryButtonComponent
import com.appsamurai.storyly.storybar.ui.model.StoryCommentComponent
import com.appsamurai.storyly.storybar.ui.model.StoryEmojiComponent
import com.appsamurai.storyly.storybar.ui.model.StoryGroup
import com.appsamurai.storyly.storybar.ui.model.StoryGroupBadgeStyle
import com.appsamurai.storyly.storybar.ui.model.StoryGroupStyle
import com.appsamurai.storyly.storybar.ui.model.StoryImageQuizComponent
import com.appsamurai.storyly.storybar.ui.model.StoryPollComponent
import com.appsamurai.storyly.storybar.ui.model.StoryProductCardComponent
import com.appsamurai.storyly.storybar.ui.model.StoryProductCatalogComponent
import com.appsamurai.storyly.storybar.ui.model.StoryProductTagComponent
import com.appsamurai.storyly.storybar.ui.model.StoryPromoCodeComponent
import com.appsamurai.storyly.storybar.ui.model.StoryQuizComponent
import com.appsamurai.storyly.storybar.ui.model.StoryRatingComponent
import com.appsamurai.storyly.storybar.ui.model.StorySwipeComponent
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.encodeSTRProductItem
import com.appsamurai.storyly.storyly_placement_flutter.common.data.util.toHexString


fun encodeStoryBarDataPayload(data: StoryBarDataPayload): Map<String, Any?> {
    return mapOf(
        "type" to STRWidgetType.StoryBar.raw,
        "items" to data.items.map { encodeStoryGroupItem(it) }
    )
}

fun encodeStoryBarPayload(payload: StoryBarPayload?): Map<String, Any?>? {
    payload ?: return null
    return mapOf(
        "component" to encodeStoryBarComponent(payload.component),
        "group" to encodeStoryGroupItem(payload.group),
        "story" to encodeStory(payload.story),
    )
}

fun encodeStoryGroupItem(group: StoryGroup?): Map<String, Any?>? {
    group ?: return null
    return mapOf(
        "id" to group.uniqueId,
        "title" to group.title,
        "name" to group.name,
        "index" to group.index,
        "seen" to group.seen,
        "iconUrl" to group.iconUrl,
        "pinned" to group.pinned,
        "type" to group.type.customName,
        "nudge" to group.nudge,
        "iconVideoUrl" to group.iconVideoUrl,
        "iconVideoThumbnailUrl" to group.iconVideoThumbnailUrl,
        "style" to encodeStoryGroupStyle(group.style),
        "stories" to group.stories.map { story -> encodeStory(story) }
    )
}

fun encodeStoryGroupStyle(style: StoryGroupStyle?): Map<String, Any?>? {
    style ?: return null
    return mapOf(
        "badge" to encodeStoryGroupBadgeStyle(style.badge),
        "borderUnseenColors" to style.borderUnseenColors?.map { it.toHexString() },
        "textUnseenColor" to style.textUnseenColor?.toHexString(),
    )
}

fun encodeStoryGroupBadgeStyle(badgeStyle: StoryGroupBadgeStyle?): Map<String, Any?>? {
    badgeStyle ?: return null
    return mapOf(
        "text" to badgeStyle.text,
        "textColor" to badgeStyle.textColor?.toHexString(),
        "backgroundColor" to badgeStyle.backgroundColor?.toHexString(),
        "endTime" to badgeStyle.endTime,
        "template" to badgeStyle.template,
    )
}

fun encodeStory(story: Story?): Map<String, Any?>? {
    story ?: return null
    return mapOf(
        "id" to story.uniqueId,
        "title" to story.title,
        "name" to story.name,
        "index" to story.index,
        "seen" to story.seen,
        "previewUrl" to story.previewUrl,
        "actionUrl" to story.actionUrl,
        "actionProducts" to story.actionProducts?.map { encodeSTRProductItem(it) },
        "currentTime" to story.currentTime,
        "storyComponentList" to story.storyComponentList?.map { encodeStoryBarComponent(it) }
    )
}

fun encodeStoryBarComponent(component: StoryBarComponent?): Map<String, Any?>? {
    component ?: return null
    val base = mapOf(
        "id" to component.id,
        "type" to component.type.get(),
        "customPayload" to component.customPayload,
    )
    
    val additional = when(component) {
        is StoryQuizComponent -> mapOf(
            "title" to component.title,
            "options" to component.options,
            "rightAnswerIndex" to component.rightAnswerIndex,
            "selectedOptionIndex" to component.selectedOptionIndex,
        )
        is StoryImageQuizComponent -> mapOf(
            "title" to component.title,
            "options" to component.options,
            "rightAnswerIndex" to component.rightAnswerIndex,
            "selectedOptionIndex" to component.selectedOptionIndex,
        )
        is StoryPollComponent -> mapOf(
            "title" to component.title,
            "options" to component.options,
            "selectedOptionIndex" to component.selectedOptionIndex,
        )
        is StoryEmojiComponent -> mapOf(
            "emojiCodes" to component.emojiCodes,
            "selectedEmojiIndex" to component.selectedEmojiIndex,
        )
        is StoryRatingComponent -> mapOf(
            "emojiCode" to component.emojiCode,
            "rating" to component.rating,
        )
        is StoryPromoCodeComponent -> mapOf(
            "text" to component.text,
        )
        is StoryCommentComponent -> mapOf(
            "text" to component.text,
        )
        is StorySwipeComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is StoryButtonComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is StoryProductTagComponent -> mapOf(
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is StoryProductCardComponent -> mapOf(
            "text" to component.text,
            "actionUrl" to component.actionUrl,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        is StoryProductCatalogComponent -> mapOf(
            "actionUrlList" to component.actionUrlList,
            "products" to component.products?.map { encodeSTRProductItem(it) },
        )
        else -> emptyMap()
    }
    
    return base + additional
}